src/openrct2-ui/windows/Ride.cpp

```diff
@@ -2223,6 +2238,115 @@ static void window_ride_show_ride_type_dropdown(rct_window *w, rct_widget *widge
+    {
+        selectionShouldBeExpanded = true;
+        rideTypeIterator = 0;
+        rideTypeIteratorMax = RIDE_TYPE_COUNT - 1;
+    }
+    else
+    {
+        selectionShouldBeExpanded = false;
+        rideTypeIterator = ride->type;
+        rideTypeIteratorMax = ride->type;
+    }
+
+    // Don't repopulate the list if we just did.
+    auto& ls = OpenRCT2::GetContext()->GetLocalisationService();
+    if (VehicleDropdownExpanded == selectionShouldBeExpanded &&
+        VehicleDropdownRideType == rideEntry &&
+        VehicleDropdownDataLanguage == ls.GetCurrentLanguage())
+        return;
+
+    VehicleDropdownData.clear();
+
+    for (; rideTypeIterator <= rideTypeIteratorMax; rideTypeIterator++)
+    {
+        if (selectionShouldBeExpanded && ride_type_has_flag(rideTypeIterator, RIDE_TYPE_FLAG_FLAT_RIDE))
+            continue;
+        if (selectionShouldBeExpanded && (rideTypeIterator == RIDE_TYPE_MAZE || rideTypeIterator == RIDE_TYPE_MINI_GOLF))
+            continue;
+
+        uint8 *rideEntryIndexPtr = get_ride_entry_indices_for_ride_type(rideTypeIterator);
+
+        for (uint8 *currentRideEntryIndex = rideEntryIndexPtr; *currentRideEntryIndex != RIDE_ENTRY_INDEX_NULL; currentRideEntryIndex++)
+        {
+            sint32 rideEntryIndex = *currentRideEntryIndex;
+            rct_ride_entry *currentRideEntry = get_ride_entry(rideEntryIndex);
+
+            // Skip if vehicle type has not been invented yet
+            if (!ride_entry_is_invented(rideEntryIndex) && !gCheatsIgnoreResearchStatus)
+                continue;
+
+            // Skip if vehicle does not belong to the same ride group
+            if (RideGroupManager::RideTypeHasRideGroups(ride->type) && !selectionShouldBeExpanded)
+            {
+                const RideGroup *rideGroup = RideGroupManager::GetRideGroup(ride->type, rideEntry);
+                const RideGroup *currentRideGroup = RideGroupManager::GetRideGroup(ride->type, currentRideEntry);
+
+                if (!rideGroup->Equals(currentRideGroup))
+                    continue;
+            }
+
+            VehicleTypeLabel label = { rideEntryIndex, currentRideEntry->naming.name, ls.GetString(currentRideEntry->naming.name) };
+            VehicleDropdownData.push_back(label);
+        }
+    }
+
+    std::sort(VehicleDropdownData.begin(), VehicleDropdownData.end(), [](auto& a, auto& b) { return std::strcmp(a.label_string, b.label_string) < 0; });
+
+    VehicleDropdownExpanded = selectionShouldBeExpanded;
+    VehicleDropdownRideType = rideEntry;
+    VehicleDropdownDataLanguage = ls.GetCurrentLanguage();
+}
+
+static void window_ride_show_vehicle_type_dropdown(rct_window *w, rct_widget *widget)
+{
+    Ride *ride = get_ride(w->number);
+    populate_vehicle_type_dropdown(ride);
+
+    size_t numItems = std::min<size_t>(VehicleDropdownData.size(), DROPDOWN_ITEMS_MAX_SIZE);
```

Comment:
`warning C4267: 'argument': conversion from 'size_t' to 'const uint8', possible loss of data`